gh-132314: fix stack array init warning

dura0ok opened the pull request:
Dummy warning fix

<!-- gh-issue-number: gh-132314 -->
* Issue: gh-132314
<!-- /gh-issue-number -->

chris-eibl commented on Python/ceval.c:
```diff
@@ -1814,7 +1814,7 @@ _PyEvalFramePushAndInit_Ex(PyThreadState *tstate, _PyStackRef func,
     _PyStackRef *newargs;
     PyObject *const *object_array = NULL;
-    _PyStackRef stack_array[8];
+    _PyStackRef stack_array[8] = {};
```
IMHO, the compiler warning is a false-positive.

If the `stack_array` is used
https://github.com/python/cpython/blob/ad3bbe8fbcc9045c436b0c5e49b1aec018c59c2d/Python/ceval.c#L1832-L1834
it is initialized

https://github.com/python/cpython/blob/ad3bbe8fbcc9045c436b0c5e49b1aec018c59c2d/Python/ceval.c#L1844-L1846

Having a small array on the stack to avoid the `PyMem_Malloc` is a common pattern, e.g. further down
https://github.com/python/cpython/blob/ad3bbe8fbcc9045c436b0c5e49b1aec018c59c2d/Python/ceval.c#L1880
or like mentioned in the issue https://github.com/python/cpython/issues/132314
https://github.com/python/cpython/blob/ad3bbe8fbcc9045c436b0c5e49b1aec018c59c2d/Python/bltinmodule.c#L1450

They all follow the same pattern, and if the compiler is not smart enough to detect it, it will issue a warning.

IMHO, initializing those stack variables using `= {}` is a waste of CPU cycles, and we should silence the warning using a [diagnostic pragma](https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html).

dura0ok replied:
I think hiding warnings using pragma it is not good idea, because we have other compilers.

chris-eibl replied:
Yeah, but not all of them emit the warning, so we only have to silence those that do?

picnixz replied:
Yes and we can selectively decide which compilers need to be silenced as well so I think for a falsey positive like that it's better to silence the warning.

dura0ok replied:
Fixed. Check please new version. @chris-eibl @picnixz

chris-eibl replied:
Sorry @dura0ok, I totally missed this one - but https://github.com/python/cpython/pull/134207 rang a bell and so I am here again.

I had a simpler approach in mind just using a pragma for GCC directly in the affected places but I like your approach, too, since it is re-usable.

There is already infrastructure and I'd place the macro next to
https://github.com/python/cpython/blob/c740fe3bd092911d9e474bcc0eed2a009482be9f/Include/pyport.h#L297-L302
I'd also go with the naming scheme used there, e.g. `_Py_COMP_DIAG_IGNORE_MAYBE_UNINITIALIZED`, and use it like
https://github.com/python/cpython/blob/c740fe3bd092911d9e474bcc0eed2a009482be9f/Modules/main.c#L522-L530

dura0ok committed:
Suppress uninitialized variable warnings for small stack allocations

Add macros to disable and re-enable compiler-specific warnings about
possibly uninitialized variables (`-Wmaybe-uninitialized` for GCC,
`-Wuninitialized` for Clang, and warning C4700 for MSVC). Use these
macros in `_PyEvalFramePushAndInit_Ex()` to suppress false positives
on stack-allocated arrays like `stack_array`.

This also addresses warnings such as the one in `pycore_call.h` when
using `small_stack[_PY_FASTCALL_SMALL_STACK]` that may be flagged by
GCC's `-Wmaybe-uninitialized`.
